drivers: power: lt3074: add driver for LT3074

cencarna opened the pull request:
## Pull Request Description

The LT3074 is a low voltage, ultralow noise, and ultrafast transient
response linear regulator with a PMBus serial interface. The device
supplies up to 3A with a typical dropout voltage of 45mV.

This pull request adds the no-OS and IIO driver, as well as examples
and documentations, to quickly evaluate the LT3074.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,706 @@
+ * @return ret    - Result of the reading procedure.
+ * 		    In case of success, the size of the read data is returned.
+*/
+static int32_t lt3074_iio_reg_read(void *dev, uint32_t reg, uint32_t *readval)
```
actually there is one comment: the return type of the functions should be `int`.

cencarna replied:
These are mapped to `.debug_reg_read` and `.debug_reg_write` which return `int32_t`. Compiler also gives warnings about incompatible pointers. Should I still proceed with `int`?

amiclaus replied:
It's fine like this, i did not take into account that aspect.

CiprianRegus commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,706 @@
+	case LT3074_READ_TEMPERATURE_1:
+	case LT3074_IC_DEVICE_REV:
+	case LT3074_MFR_SPECIAL_ID:
+		return lt3074_read_word(lt3074, (uint8_t)reg,
+					(uint16_t *)readval);
```
Don't cast pointers. The result will be different based on CPU endianess

cencarna commented:
v2:
- Removed pointer casts. Used intermediate variables instead to handle type differences.

buha reviewed (changes requested):
Please also update the project to the new project infrastructure.

Here's the PR with the changes with 3 modified projects as example: https://github.com/analogdevicesinc/no-OS/pull/2412

And here's a pending PR where I gave feedback with pretty much that needs to be changed, same applies to your PR.
https://github.com/analogdevicesinc/no-OS/pull/2411

CiprianRegus commented on drivers/power/lt3074/lt3074.c:
```diff
@@ -0,0 +1,778 @@
+	}
+
+	mant = no_os_clamp(NO_OS_DIV_ROUND_CLOSEST_ULL(data, MILLI),
+			   0, 0x3FF);
```
`NO_OS_GENMASK()`

CiprianRegus commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,727 @@
+	if (!iio_desc)
+		return -ENODEV;
+
+	no_os_free(iio_desc->iio_dev->channels);
```
The `lt3074_channels` array is statically allocated.

CiprianRegus commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,727 @@
+	if (ret)
+		return ret;
+
+	vals[0] = (int32_t)(word / MILLI);
+	vals[1] = (int32_t)((word * MILLI) % MICRO);
```
These are implicit casts, so they don't have any effect.

buha commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,727 @@
+	case LT3074_IOUT_OC_FAULT_RESPONSE:
+	case LT3074_REVISION:
+	case LT3074_MFR_DEFAULT_CONFIG:
+		ret = lt3074_read_byte(lt3074, (uint8_t)reg, &byte);
```
this will definitely work but i expect this kind of register level access size to be handled in the device driver so at the lower level, not here

buha replied:
this is how we prefer to do it:
https://github.com/analogdevicesinc/no-OS/blob/2d0e5d5ed9244bf91e4b47870702a696406aab71/drivers/adc/ad9250/ad9250.h#L54

each register has size information coded into the register "address" macro so the lowest level spi/i2c/whatever read/write, knows how many bytes to read/write

buha commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,727 @@
+static struct iio_attribute lt3074_global_attrs[] = {
+	{
+		.name = "vout_margin_low",
+		.show = lt3074_iio_read_vout_margin,
```
please use a single function for all of these if they are similar enough to avoid code duplication

buha commented on projects/lt3074/Makefile:
```diff
@@ -0,0 +1,9 @@
+# Select the example you want to enable by choosing y for enabling and n for disabling
+BASIC_EXAMPLE = n
```
and i see you didn't yet update the project structure, please update it

cencarna commented:
v3:

- Remove `free` for statically allocated `lt3074_channels` array
- Used `NO_OS_GENMASK` for mantissa computation
- Removed redundant typecasts for variable assignments that are already implicitly cast
- Moved all register-level access to device driver. 
- Macros now have a prefix for size information
- Generic reg read and reg write for size-encoded macros
- Used single function for all attribute reads and writes
- Updated project structure using examples.mk

CiprianRegus commented on drivers/power/lt3074/lt3074.h:
```diff
@@ -0,0 +1,275 @@
+#define LT3074_STATUS_ALL_TYPE_MSK		0xFF
+
+/* LINEAR data format params */
+#define LT3074_LIN11_MANTISSA_MAX		1023000L
+#define LT3074_LIN11_MANTISSA_MIN		511000L
+#define LT3074_LIN11_EXPONENT_MAX		15
+#define LT3074_LIN11_EXPONENT_MIN		-15
+#define LT3074_LIN11_MANTISSA_MSK		NO_OS_GENMASK(10,0)
+#define LT3074_LIN11_EXPONENT_MSK		NO_OS_GENMASK(15,11)
+#define LT3074_LIN11_EXPONENT(x)		((int16_t)(x) >> 11)
+#define LT3074_LIN11_MANTISSA(x)		(((int16_t)((x & 0x7FF) << 5)) >> 5)
+#define LT3074_LIN16_EXPONENT			-13
```
These could be moved to the source file, since they're not meant to be used directly

CiprianRegus commented on drivers/power/lt3074/iio_lt3074.c:
```diff
@@ -0,0 +1,456 @@
+#include "lt3074.h"
+#include "iio_lt3074.h"
+
+static const int32_t lt3074_margin_avail[] = {
+	0,
+	1,
+	3,
+	5,
+	10,
+	15,
+	20,
+	25,
+	30,
+};
```
Define these in one line.

CiprianRegus reviewed (approved):
Some minor remarks, looks good otherwise

cencarna commented:
v4:
- Moved LINEAR reg2data macros to source file
- Define `lt3074_margin_avail` in one line

kister-jimenez commented:
lgtm
